keras_nlp/models/albert/albert_backbone.py

```diff
@@ -0,0 +1,264 @@
+        )
+        padding_mask = keras.Input(
+            shape=(None,), dtype="int32", name="padding_mask"
+        )
+
+        # Embed tokens, positions, and segment ids.
+        token_embedding_layer = keras.layers.Embedding(
+            input_dim=vocabulary_size,
+            output_dim=embedding_dim,
+            embeddings_initializer=albert_kernel_initializer(),
+            name="token_embedding",
+        )
+        token_embedding = token_embedding_layer(token_id_input)
+        position_embedding = PositionEmbedding(
+            initializer=albert_kernel_initializer(),
+            sequence_length=max_sequence_length,
+            name="position_embedding",
+        )(token_embedding)
+        segment_embedding = keras.layers.Embedding(
+            input_dim=num_segments,
+            output_dim=embedding_dim,
+            embeddings_initializer=albert_kernel_initializer(),
+            name="segment_embedding",
+        )(segment_id_input)
+
+        # Sum, normalize and apply dropout to embeddings.
+        x = keras.layers.Add()(
+            (token_embedding, position_embedding, segment_embedding)
+        )
+        x = keras.layers.LayerNormalization(
+            name="embeddings_layer_norm",
+            axis=-1,
+            epsilon=1e-12,
+            dtype=tf.float32,
+        )(x)
+        x = keras.layers.Dropout(
+            dropout,
+            name="embeddings_dropout",
+        )(x)
+
+        # Project the embedding to `hidden_dim`.
+        x = keras.layers.Dense(
+            hidden_dim,
+            kernel_initializer=albert_kernel_initializer(),
+            name="embedding_projection",
+        )(x)
+
+        def get_group_layer(group_idx):
+            """Defines a group `num_inner_repetitions` transformer layers and
+            returns the callable.
+            """
+            transformer_layers = [
+                TransformerEncoder(
+                    num_heads=num_heads,
+                    intermediate_dim=intermediate_dim,
+                    activation=lambda x: keras.activations.gelu(
+                        x, approximate=True
+                    ),
+                    dropout=dropout,
+                    kernel_initializer=albert_kernel_initializer(),
+                    name=f"group_{group_idx}_transformer_layer_{inner_repetition_idx}",
+                )
+                for inner_repetition_idx in range(num_inner_repetitions)
+            ]
+
+            def call(x, padding_mask):
+                for transformer_layer in transformer_layers:
+                    x = transformer_layer(x, padding_mask=padding_mask)
+                return x
+
+            return call
+
+        layer_idx = 0
+        # Note: `num_calls_per_group` is sort of a misnomer; it is not an
+        # integer, and every group may not be called the same number of times.
+        num_calls_per_group = num_layers / num_groups
+        for group_idx in range(num_groups):
```

Comment:
This still seems a bit instructable to me. What do you think of this approach?

```python
max_calls_per_group = num_layers / num_groups
group_idxs = [idx for idx in range(num_groups) for k in range(max_calls_per_group)][:num_layers]
group_layers = [get_group_layer(idx) for idx in range(num_groups)]
for group_idx in group_idxs:
    x = group_layers[group_idx](x, padding_mask=padding_mask)
```

Reply:
Hmmm, I don't think it will work.

```
import math
num_layers = 8
num_groups = 5
max_calls_per_group = math.ceil(num_layers / num_groups)
# Repeat each group id `max_calls_per_group` times and truncate at `num_layers`
group_idxs = [idx for idx in range(num_groups) for k in range(max_calls_per_group)][:num_layers]
print(group_idxs)
```

gives 

```
[0, 0, 1, 1, 2, 2, 3, 3]
```

which is different from 

```
 0, 0, 1, 1, 2, 3, 3, 4
```

Check this comment out: https://github.com/keras-team/keras-nlp/pull/622#discussion_r1063184019...explains this whole thing the best.

Reply:
I see that @mattdangerw also proposed something similar above, sorry for missing that.

To me this seems horribly overengineered and confusing, and all the back and forth attests to that. I think we could load the existing checkpoints without replicating this logic blow-for-blow---they only have one unique layer, right!? The only advantage would be if a user wants a more complicated custom configuration...but at that point would they care if replicated the paper logic vs something simpler?

Reply:
Hmmm, yeah! Well, dunno. The logic is sort of weird, but I feel like this is the only "salient" feature of ALBERT, so to speak. Anyway, will do what you guys think is right!

Reply:
The alternative is the enforce this restriction that `num_layers` should always be divisible by `num_groups`. We can make the logic simpler, @jbischof's snippet will work in that case!

Reply:
I like that restriction! Let's do it!